File: reverse-vowels-of-a-string/reverse-vowels-of-a-string.java
class Solution {
    List<Character> vowels = Arrays.asList('a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U');
    public String reverseVowels(String s) {
        Deque<Character> stack = new ArrayDeque<>();
        char[] reversed = new char[s.length()];
        for (int i = 0; i < s.length(); i++) {
            if (!vowels.contains(s.charAt(i))) {
                reversed[i] = s.charAt(i);
            }
            else {
                stack.offer(s.charAt(i));
            }
        }
        while (!stack.isEmpty()) {
            for (int i = 0; i < s.length(); i++) {
                if (reversed[i] == 0) {
                    reversed[i] = stack.pollLast();
                }
            }
        }
        return String.valueOf(reversed);
    }
}

File: 345-reverse-vowels-of-a-string/reverse-vowels-of-a-string.py
class Solution:
    def reverseVowels(self, s: str) -> str:
        stack = deque()
        string = list(s)
        for i, char in enumerate(string):
            if char.lower() in ('a', 'e', 'i', 'o', 'u'):
                stack.append(char)
                string[i] = '_'
        for i, char in enumerate(string):
            if char == '_':
                string[i] = stack.pop()
        return ''.join(string)